PR: Add position to text completion

impact27 opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)
* [ ] Included a screenshot (if affecting the UI)

<!--- Explain what you've done and why --->
While an autocomplete is being processed, the cursor can move. Auto completing on another line is clearly not what is expected. The cursor is therefore moved back if this happens. 
Also, the selection of an option in the list was broken.



### Issue(s) Resolved

<!--- List the issue(s) below, in the form "Fixes #1234"; one per line --->

Fixes #8727 
Fixes #8723 


### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct:
Impact27
<!--- Thanks for your help making Spyder better for everyone! --->

pep8speaks commented:
Hello @impact27! Thanks for updating the PR.





 - In the file [`spyder/utils/icon_manager.py`](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py), following are the PEP8 issues :

> [Line 288:80](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py#L288): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (109 > 79 characters)
> [Line 289:80](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py#L289): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (109 > 79 characters)
> [Line 290:63](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py#L290): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 290:80](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py#L290): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 79 characters)
> [Line 291:72](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py#L291): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 291:80](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py#L291): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (119 > 79 characters)
> [Line 292:72](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py#L292): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 292:80](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py#L292): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (119 > 79 characters)
> [Line 293:63](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py#L293): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 293:80](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py#L293): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 79 characters)
> [Line 294:63](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py#L294): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 294:80](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py#L294): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 79 characters)
> [Line 295:60](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py#L295): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 295:80](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py#L295): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 79 characters)
> [Line 296:59](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py#L296): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 296:80](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py#L296): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 79 characters)
> [Line 297:80](https://github.com/spyder-ide/spyder/blob/430d1e541add6c03b9fa09159627bc30c29a0dbe/spyder/utils/icon_manager.py#L297): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 79 characters)

ccordoba12 commented on spyder/plugins/editor/widgets/base.py:
```diff
@@ -183,16 +184,30 @@ def keyPressEvent(self, event):
             self.hide()
             QListWidget.keyPressEvent(self, event)
 
+    def sendKeyToTextEdit(self, event):
```
Please change this method's name to `send_key_to_textedit`.

ccordoba12 commented on spyder/plugins/editor/widgets/base.py:
```diff
@@ -183,16 +184,30 @@ def keyPressEvent(self, event):
 
+    def sendKeyToTextEdit(self, event):
+        """Send key to text edit and updates the position of
+        the text completion"""
```
Please rewrite this to

```
"""
Send key to text edit and updates the position of
the text completion.
"""
```

impact27 commented:
I changed it a bit: the sendKeyToTextEdit didn't really make sense because the cursor should not be before the position where you started autocompleting. Therefore I changed it so that it closes if you press backspace.

impact27 commented:
Anyway that makes more sense, if I write:
`np.newb`
and I want `np.newaxis`, if I press backspace I will want a new request anyway. (So new tab)

impact27 commented:
One issue that is still there is:
`import numpy as np\nnp.ne<tab><backspace>\n`
and then `<enter>` when the list appears.
This will place the `negative` on the next line (because the position changed as a letter was removed).
I am not sure it is a big issue as that would only apply if somebody messes with the characters before the completion position In the time between the `<tab>` and the reply from LSP.

impact27 commented:
I would need some help with the tests. 
what I want to test is:
 1. Complete `math.a <tab>` ... `s <enter>` to `math.asin`
This will check if asin can be selected in the list
 2. Complete `math.a <tab><enter>` ... `<enter>` to `math.acos`
This will check if when enter is pressed too rapidly, the correction is still correct
 3. Complete` math.a <tab> s` ...`<enter>` to `math.asin`
This will check if when s is pressed too rapidly, asin is still selected

impact27 commented:
I fixed the tests

ccordoba12 commented:
Thanks! This looks really good but I'm going to ask @dalthviz to give it a final review since he has worked on code completion a lot.

dalthviz commented:
Hi @impact27 thanks for working in this! While reviewing I encountered an error:

```python
Traceback (most recent call last):
  File "C:\Users\Daniel\Desktop\spyder_others\spyder-impact\spyder\spyder\plugins\editor\lsp\client.py", line 230, in on_msg_received
    handler(resp['result'], req_id)
  File "C:\Users\Daniel\Desktop\spyder_others\spyder-impact\spyder\spyder\plugins\editor\lsp\providers\document.py", line 117, in process_document_completion
    LSPRequestTypes.DOCUMENT_COMPLETION, {'params': response})
  File "C:\Users\Daniel\Desktop\spyder_others\spyder-impact\spyder\spyder\plugins\editor\widgets\codeeditor.py", line 690, in handle_response
    handler(params)
  File "C:\Users\Daniel\Desktop\spyder_others\spyder-impact\spyder\spyder\plugins\editor\widgets\codeeditor.py", line 791, in process_completion
    self.completion_widget.show_list(completion_list, position)
  File "C:\Users\Daniel\Desktop\spyder_others\spyder-impact\spyder\spyder\plugins\editor\widgets\base.py", line 64, in show_list
    if self.textedit.textCursor().position() < position:
TypeError: '<' not supported between instances of 'int' and 'NoneType'
```

I was trying to complete `import num -> import numpy` 

The error appeared only the first time I tried (I'm unsure of how `position` could had a `None` value). After restarting Spyder it worked without issue. Could you check what is happening or maybe add a validation to prevent position being `None`? Thanks :)

impact27 commented:
I have added a validation. Maybe the function is called twice?

impact27 commented:
<tab><tab> does the trick

ccordoba12 commented:
@impact27, please merge with master to get the fixes to our tests I pushed yesterday.

ccordoba12 commented:
Thanks @impact27 for your work and @dalthviz for the review!
